ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/WebSocketDeflateExtension.kt

```diff
@@ -0,0 +1,209 @@
+        val protocol = negotiatedProtocols.find { it.name == PERMESSAGE_DEFLATE } ?: return false
+
+        incomingContextTakeover = config.serverNoContextTakeOver
+        outgoingContextTakeover = config.clientNoContextTakeOver
+
+        for ((key, value) in protocol.parseParameters()) {
+            when (key) {
+                SERVER_MAX_WINDOW_BITS -> {
+                    // This value is a hint for a client and can be ignored.
+                }
+                CLIENT_MAX_WINDOW_BITS -> {
+                    if (value.isBlank()) continue
+                    check(value.toInt() == MAX_WINDOW_BITS) { "Only $MAX_WINDOW_BITS window size is supported." }
+                }
+                SERVER_NO_CONTEXT_TAKEOVER -> {
+                    check(value.isBlank())
+                }
+                CLIENT_NO_CONTEXT_TAKEOVER -> {
+                    check(value.isBlank())
+                }
+            }
+        }
+
+        return true
+    }
+
+    override fun serverNegotiation(requestedProtocols: List<WebSocketExtensionHeader>): List<WebSocketExtensionHeader> {
+        val protocol = requestedProtocols.find { it.name == PERMESSAGE_DEFLATE } ?: return emptyList()
+        val parameters = mutableListOf<String>()
+
+        for ((key, value) in protocol.parseParameters()) {
+            when (key.toLowerCase()) {
+                SERVER_MAX_WINDOW_BITS -> {
+                    check(value.toInt() == MAX_WINDOW_BITS) { "Only $MAX_WINDOW_BITS window size is supported" }
+                }
+                CLIENT_MAX_WINDOW_BITS -> {
+                    // This value is a hint for a server and can be ignored.
+                }
+                SERVER_NO_CONTEXT_TAKEOVER -> {
+                    check(value.isBlank())
+
+                    outgoingContextTakeover = false
+                    parameters.add(SERVER_NO_CONTEXT_TAKEOVER)
+                }
+                CLIENT_NO_CONTEXT_TAKEOVER -> {
+                    check(value.isBlank())
+
+                    incomingContextTakeover = false
+                    parameters.add(CLIENT_NO_CONTEXT_TAKEOVER)
+                }
+                else -> error("Unsupported extension parameter: ($key, $value)")
+            }
+        }
+
+        return listOf(WebSocketExtensionHeader(PERMESSAGE_DEFLATE, parameters))
+    }
+
+    override fun processOutgoingFrame(frame: Frame): Frame {
+        if (frame !is Frame.Text && frame !is Frame.Binary) return frame
+
+        val deflated = deflater.deflateFully(frame.data)
+
+        if (!outgoingContextTakeover) {
+            deflater.reset()
+        }
+
+        return Frame.byType(frame.fin, frame.frameType, deflated, rsv1, frame.rsv2, frame.rsv3)
+    }
+
+    override fun processIncomingFrame(frame: Frame): Frame {
+        if (!frame.rsv1 || (frame !is Frame.Text && frame !is Frame.Binary)) return frame
+
+        val inflated = inflater.inflateFully(frame.data)
+
+        if (!outgoingContextTakeover) {
```

Comment:
incomingContextTakeover instead? Looks like we don't have enough tests for these rear options...

Reply:
Yep, sorry not tested